File: apps/jetbrains/plugin/src/main/kotlin/io/autocomplete/bridge/SettingsExport.kt
package io.autocomplete.bridge

/**
 * Produces the portable settings export shared by the JCEF bridge.
 *
 * API-key state and arbitrary extra headers must never leave the local host:
 * header values can carry credentials even when they are not named "apiKey".
 */
object SettingsExport {
    fun redact(snapshot: Map<String, Any?>): Map<String, Any?> {
        val profiles = snapshot["profiles"] as? List<*> ?: return snapshot
        val sanitizedProfiles =
            profiles.mapNotNull { item ->
                val profile = item as? Map<*, *> ?: return@mapNotNull null
                profile.entries
                    .associate { it.key.toString() to it.value }
                    .filterKeys { it !in EXCLUDED_PROFILE_FIELDS }
            }
        return snapshot + ("profiles" to sanitizedProfiles)
    }

    private val EXCLUDED_PROFILE_FIELDS =
        setOf("apiKey", "hasApiKey", "authHeaderTemplate", "extraHeadersJson")
}

File: apps/vscode/extension/src/config/settingsExport.ts
/**
 * Removes profile fields that must never leave the current host.
 *
 * `extraHeadersJson` stays available in the in-process UI snapshot so the user
 * can edit non-secret routing headers, but it is intentionally non-portable:
 * arbitrary header values can contain credentials and are omitted from export.
 */
export function redactSettingsExport<T extends { profiles: readonly object[] }>(snapshot: T) {
  return {
    ...snapshot,
    profiles: snapshot.profiles.map((profile) => {
      const {
        apiKey: _apiKey,
        hasApiKey: _hasApiKey,
        authHeaderTemplate: _authHeaderTemplate,
        extraHeadersJson: _extraHeadersJson,
        ...safe
      } = profile as object & {
          apiKey?: unknown;
          hasApiKey?: unknown;
          authHeaderTemplate?: unknown;
          extraHeadersJson?: unknown;
        };
      return safe;
    }),
  };
}
